File: algorithms/6. ZigZag Conversion/solution.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows == 1 or numRows >= len(s):
            return s
        matrix, index, direction = [''] * numRows, 0, 1
        for c in s:
            matrix[index] += c
            index += direction
            if index in (0, numRows - 1):
                direction *= -1
        return ''.join(matrix)

# 1158 / 1158 test cases passed.
# Status: Accepted
# Runtime: 112 ms
# beats 74.57 %

File: algorithms/6. ZigZag Conversion/solution.cc
class Solution {
public:
    string convert(string s, int numRows) {
        if ( numRows < 2 || numRows >= s.size() )
            return s;
        vector<string> matrix(numRows);
        int row = 0, direction = 1;
        for ( char c : s ) {
            matrix[row].push_back(c);
            row += direction;
            if ( 0 == row || numRows - 1 == row )
                direction *= -1;
        }
        string result;
        for ( string & r : matrix )
            result += r;
        return result;
    }
};

// 1158 / 1158 test cases passed.
// Status: Accepted
// Runtime: 25 ms
// beats 72.41 %
